contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1683 @@
+        require(existingNode != address(0), ClusterLib.CANNOT_BE_ADDRESS_ZERO);
+        if (nodes[existingNode].node == address(0)) {
+            revert(ClusterLib.NODE_NOT_PRESENT_IN_THE_TREE);
+        } else {
+            int256 differenceInKeyBalance = int256(newBalance) - int256(nodes[existingNode].balance);
+            _update(root, existingNode, differenceInKeyBalance);
+        }
+    }
+
+    /// @notice Search a single node from the tree. Probability of getting selected is proportional to node's balance
+    /// @param randomizer random number used for traversing the tree
+    /// @return Address of the selected node
+    function weightedSearch(uint256 randomizer) public view returns (address) {
+        uint256 totalWeightInTree = _getTotalBalancesIncludingWeight(root);
+        uint256 searchNumber = randomizer % totalWeightInTree;
+        // console2.log("totalWeightInTree", totalWeightInTree);
+        // console2.log("searchNumber", searchNumber);
+        return _weightedSearch(root, searchNumber);
+    }
+
+    /// @notice internal function to recursively search the node
+    /// @param _node address of the node
+    /// @param searchNumber random number used for traversing the tree
+    /// @return Address of the selected node
+    function _weightedSearch(address _node, uint256 searchNumber) public view returns (address) {
+        // |-----------sumOfLeftWeight -------|----balance-----|------sumOfRightWeights------|
+        Node memory node = nodes[_node];
+        (uint256 index1, uint256 index2, uint256 index3) = _getIndexes(_node);
+
+        if (searchNumber <= index1) {
+            return _weightedSearch(node.left, searchNumber);
+        } else if (searchNumber > index1 && searchNumber <= index2) {
+            return _node;
+        } else if (searchNumber > index2 && searchNumber <= index3) {
+            return _weightedSearch(node.right, searchNumber - index2);
+        } else {
+            // _printNode(_node);
+            // console2.log("indexes", index1, index2, index3);
+            // console2.log("search number", searchNumber);
+            revert(ClusterLib.ERROR_OCCURED_DURING_WEIGHTED_SEARCH);
+        }
+    }
+
+    /// @notice Returns the indexes for the node which will be used to traverse the tree
+    /// @param _node Address of the node
+    function _getIndexes(address _node)
+        internal
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        Node memory node = nodes[_node];
+        return (
+            node.sumOfLeftBalances,
+            node.sumOfLeftBalances + node.balance,
+            node.sumOfLeftBalances + node.balance + node.sumOfRightBalances
+        );
+    }
+
+    /// @notice Update the balance of the node
+    /// @param node Address of the current node
+    /// @param key Address of the key
+    /// @param diff Difference in the balance of the key
+    function _update(
+        address node,
```

Comment:
`root`  might be a more appropriate name for this as it is the root of the subtree for which you are updating